### PR TITLE
[WIP] Plugins & Extensions

### LINKED ISSUE #243: Exclude extensions by id or priority

### PR DESCRIPTION
**Pull request by arturbosch**
Fixes #243 , #171, #134, #101

### CODE REVIEW

**Comment by vanniktech** [comment]
Let me know once you're done and this is no longer WIP and I'll be happy to review it.

**Comment by arturbosch** [comment]
@vanniktech @Mauin 

1. With the current design the user can generate multiple output reports at once eg. plain and xml + custom formats. This will change `--output` to be a directory rather a path to a single file.  The `--output-format` will disappear and excluding formats can be done via configuration. What do you think of this design? This will imply if the user runs detekt in cli without specifying a configuration, both 'default' output reports (xml, plain) get printed if `--output` is set.

2. The default unspecified config is `Config.empty` should we change it to the default config provided by detekt? It is the same as `--generate-config`?

**Comment by vanniktech** [comment]
1. `printed` to the console or to a file? I like that you'll get plain + xml formats.
2. Yes that would be a first step. Maybe even failFast as a default. (Although maybe I'm too opinionated there)

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Extension.kt [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package io.gitlab.arturbosch.detekt.api
+
+/**
+ * @author Artur Bosch
+ */
+interface Extension {
+	val id: String get() = javaClass.simpleName
```
Comment (vanniktech) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Extension.kt]: won't this always be `Extension` then?

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Extension.kt [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+	val priority: Int get() = -1
```
Comment (vanniktech) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Extension.kt]: `: Int` could be left out

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Extension.kt [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+	@Suppress("EmptyFunctionBlock")
+	fun init(config: Config) {
```
Comment (vanniktech) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Extension.kt]: should also be `{}` for consistency. FileProcessListener does it that way

File: detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/FilteredDetectionResult.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+	private val filteredFindings: Map<String, List<Finding>>
+
+	init {
+		filteredFindings = detektion.findings
+				.map { (key, value) -> key to baselineFacade.filter(value) }
+				.toMap()
+	}
+
+	override val findings: Map<String, List<Finding>> = filteredFindings
```
Comment (vanniktech) [inline detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/FilteredDetectionResult.kt]: return type can be omitted

File: detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/console/BuildFailureReport.kt [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.HashMap
+
+/**
+ * @author Artur Bosch
+ */
+class BuildFailureReport : ConsoleReport() {
+
+	override val priority: Int = Int.MIN_VALUE
+
+	private var buildConfig: Config by SingleAssign()
```
Comment (vanniktech) [inline detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/console/BuildFailureReport.kt]: not needed

File: detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/console/BuildFailureReport.kt [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+	private var weightsConfig: Config by SingleAssign()
+	private var warning: Int by SingleAssign()
+	private var fail: Int by SingleAssign()
+
+	override fun init(config: Config) {
+		buildConfig = config.subConfig("build")
+		weightsConfig = buildConfig.subConfig("weights")
+		warning = buildConfig.valueOrDefault("warningThreshold", -1)
+		fail = buildConfig.valueOrDefault("failThreshold", -1)
```
Comment (vanniktech) [inline detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/console/BuildFailureReport.kt]: should not the properties named the same `failThreshold`, `warningThreshold`, `weights` ?

**Comment by arturbosch** [comment]
1. To a file. `OutputReport` handles file outputs and `ConsoleReport` handles console output.

Btw the console output is now modularized. You can turn on and off different parts of the output which gets printed in M13. It is possible to turn off default findings printing and provide your own `ConsoleReport` which will do this.

2. If we use `failFast` as default then it is like having no configuration :)

File: detekt-cli/src/test/kotlin/io/gitlab/arturbosch/detekt/cli/console/BuildFailureReportSpec.kt [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+import org.jetbrains.kotlin.com.intellij.openapi.util.Key
+import org.jetbrains.spek.api.dsl.describe
+import org.jetbrains.spek.api.dsl.it
+import org.jetbrains.spek.subject.SubjectSpek
+import kotlin.test.assertFails
+
+/**
+ * @author Artur Bosch
+ */
+internal class BuildFailureReportSpec : SubjectSpek<BuildFailureReport>({
```
Comment (vanniktech) [inline detekt-cli/src/test/kotlin/io/gitlab/arturbosch/detekt/cli/console/BuildFailureReportSpec.kt]: I find it a bit confusing that some are using just `Spek` and some are using `SubjectSpek` for testing. Should not we aim for only one system / way?

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/ConsoleReport.kt [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+abstract class ConsoleReport : Extension {
+
+	fun print(printer: PrintStream, detektion: Detektion) {
+		render(detektion)?.let {
+			printer.println(it)
+		}
+	}
+
+	abstract fun render(detektion: Detektion): String?
+}
```
Comment (Mauin) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/ConsoleReport.kt]: missing newlines at the end of files

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Extension.kt (hunk 1 above)
Comment (arturbosch) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Extension.kt]: nope, javaClass gets called at runtime

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Extension.kt (hunk 2 above)
Comment (arturbosch) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Extension.kt]: In the api (interfaces at all) the return types should always be present

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Extension.kt (hunk 3 above)
Comment (arturbosch) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Extension.kt]: Intellij is inconsistent here when there is a annotation present oO

File: detekt-cli/src/test/kotlin/io/gitlab/arturbosch/detekt/cli/console/BuildFailureReportSpec.kt (hunk 7 above)
Comment (arturbosch) [inline detekt-cli/src/test/kotlin/io/gitlab/arturbosch/detekt/cli/console/BuildFailureReportSpec.kt]: If you do not need specific `TestConfig` in your rule for each test case, just use `SubjektSpek`.

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/ConsoleReport.kt (hunk 8 above)
Comment (arturbosch) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/ConsoleReport.kt]: Not in intellij :)

File: detekt-cli/src/test/kotlin/io/gitlab/arturbosch/detekt/cli/console/BuildFailureReportSpec.kt (hunk 7 above)
Comment (arturbosch) [inline detekt-cli/src/test/kotlin/io/gitlab/arturbosch/detekt/cli/console/BuildFailureReportSpec.kt]: As long as the test case is clear and clean, no worry about Spek or SubjectSpek.

**Comment by arturbosch** [comment]
I think this baby is ready to get merged :D

Should the `--rules` parameter which is now `--plugins` maybe renamed to `--extensions`?

**Comment by vanniktech** [comment]
Hmm dunno, what name describes it better?

**Comment by arturbosch** [comment]
Both names are good :D but there is now a common interface `Extension` ...

**Comment by vanniktech** [comment]
Hmm then let your gut decide.

**Pull request merged**